### ISSUE TITLE
[v2 QUESTION]: Regarding class-weight/target-weight in multitask learning

### ISSUE THREAD

**Issue by mmagithub** [chemprop/chemprop#1184]
Hi,

I am wondering if chemprop v2 has an option similar to class-weight/target-weight to compensate for data sparseness in multitask learning/training, as I cannot find this option in the documentation ?

Thanks,
Marawan

**Comment by JacksonBurns** [comment]
You can use `--class-balance` (see https://chemprop.readthedocs.io/en/latest/cmd.html) to automatically balance classification tasks during dataset splitting

**Comment by mmagithub** [comment]
How about regression ?

**Comment by KnathanM** [comment]
Maybe you are looking for `--task-weights`.

**Comment by mmagithub** [comment]
I suspected that as well, but the documentation (https://chemprop.readthedocs.io/en/latest/tutorial/python/loss_functions.html#Task-weights) doesn’t clearly indicate if this is exactly what I need. According to the documentation, task weights measure how accurate you want your predictions to be. Therefore, even if the data is balanced (non-sparse & multitask), task weights should still have an impact. Am I interpreting the documentation correctly? Additionally, an option similar to what is mentioned in this thread (https://github.com/chemprop/chemprop/issues/145) would be more helpful in my case. However, I’m not entirely sure if it’s available in Chemprop V2, as I couldn’t find any reference to it in the documentation. The alternatives are 1) create a new loss function, 2) make sure to balance the representations of the non NaNs in the mini batches.

**Comment by KnathanM** [comment]
You are right that task weights would still have an impact on non-sparse data. Task weights just bias the model towards one of the tasks for whatever reason you have. The docs suggest one reason is wanting to get one task more accurate than another. But I think another reason could be compensating for data imbalance between tasks. 

Lets say one batch of data looks like this:
|      | task1 | task2 |
|------|-------|-------|
| mol1 | 1     | 2     |
| mol2 | 3     | 4     |
| mol3 | NaN   | 5     |
| mol4 | NaN   | 6     |
| mol5 | NaN   | 7     |

We have two values for task1 and five for task2. The loss function will look like $L_{total} = L_{mol1,task1} + L_{mol1,task2} + L_{mol2,task1} + L_{mol2,task2} + L_{mol3,task2} + L_{mol4,task2} + L_{mol5,task2}$. If I understand correctly, you are worried about this being biased towards task2, so you'd want to weight the two losses for task1 by something like 2.5. Is this right?

We don't have a way to dynamically weight tasks by each batch. We can only give an overall weight to a task. I think if you have $n_{task1}$ target values and $n_{task2}$ target values, you can weight task 1 by $n_{task2}/n_{task1}$ and use a large enough batch size that with random shuffling you get on average representative batches in terms of how many values you have for each target.

**Comment by mmagithub** [comment]
Makes sense, thanks Nathan.

I also noticed that in the multi-task training documentation (https://chemprop.readthedocs.io/en/latest/tutorial/cli/train.html), there is a section that touches on this scenario:

"""
**Missing Target Values**
When training multitask models (models which predict more than one target simultaneously), sometimes not all target values are known for all molecules in the dataset. Chemprop automatically handles missing entries in the dataset by masking out the respective values in the loss function, so that partial data can be utilized.

The loss function is rescaled according to all non-missing values, and missing values do not contribute to validation or test errors. Training on partial data is therefore possible and encouraged (versus taking out datapoints with missing target entries). No keyword is needed for this behavior, it is the default.

"""

Does it conflict with using task-weight keyword ?

**Comment by KnathanM** [comment]
You could always start with weighting all tasks the same (the default option) and see if your model performance is good enough for each task. If the performance on the task with less data is poor, you could try weighting it more (by the sqrt(ratio) or something, treating it almost like a hyperparameter) and see if that helps. You may find that you just don't have enough data for that task to get a good model. 

Task weights don't conflict with masking missing values. If my targets are:
|      | task1 | task2 | task3 |
|------|-------|-------|-------|
| mol1 | t11   | t12   | t13   |
| mol2 | NaN   | t22   | t23   |
| mol3 | NaN   | NaN   | t33   |

And my predictions are 
|      | task1 | task2 | task3 |
|------|-------|-------|-------|
| mol1 | p11   | p12   | p13   |
| mol2 | p21   | p22   | p23   |
| mol3 | p31   | p32   | p33   |

And my task weights are $w_{t1}, w_{t2}, w_{t3}$ and my data point weights are $w_{d1}, w_{d2}, w_{d3}$, then my loss is calculated as
$L = \left(w_{t1}w_{d1}L(t11,p11) + w_{t2}w_{d1}L(t12,p12) + w_{t3}w_{d1}L(t13,p13) + w_{t2}w_{d2}L(t22,p22) + w_{t2}w_{d3}L(t23,p23) + w_{t3}w_{d3}L(t33,p33) \right)/6$ (6 is the number of target values I had).
